### ISSUE TITLE
Restore MUST requirement for version field first

### ISSUE THREAD

**Issue by jfirebaugh** [mapbox/vector-tile-spec#64]
At one point the spec said "The version field MUST be the first field within the layer", and as a result of #43 we changed it to SHOULD. I think this was a mistake, and we should have left it MUST, and accepted the requirement that vector tiles be generated with tools that provide control over field order. Let's consider changing it back in a future revision.

Here's the real-world scenario where it matters: for offline features in Mapbox Mobile, the SDK downloads a large number of gzipped vector tiles for offline storage. Mapbox Mobile wants to not have to support V1 tiles at some point in the future. Therefore it needs to verify that all downloaded tiles (possibly tens of thousands of them) are V2 tiles. If the version field were guaranteed to be first, it could gunzip just the first few bytes of each tile and check the version. If the version field is not guaranteed to be first, it has to either disallow otherwise-valid V2 tiles where the version field is not first, or allow an unbounded portion of the tile to be gunzipped, potentially making offline downloads much more compute intensive then they would otherwise need to be. Neither are good failure modes.

cc @flippmoke @joto

**Comment by flippmoke** [comment]
I think in a future version we should move it outside of the layer and into the Tile Message, as it should be a constant across layers as well, and then it should be required to be the first field. It could easily be field #1 in that situation and I think we would be safe to require that it is always the first field provided.

**Comment by springmeyer** [comment]
@flippmoke for tiles to be able to be concatenated there needs to be only one field (the layer) within a tile message.

**Comment by flippmoke** [comment]
@springmeyer I am not sure a simple concatenation should be our primary driver anymore here. We already have a requirement that we do not have duplicate layer names across fields and therefore simple concatenation is not possible. It would still be relatively simple to concat any additional layers to the end of the tile, you would simply strip out the version tag prior to this operation.

**Comment by jfirebaugh** [comment]
I agree that it's also worth reconsidering whether simple concatenation is a worthwhile/realistic design goal.

I didn't consider the scenario where multiple tiles of differing versions get simply concatenated. That possibility introduces additional concerns and complexities for any application that needs to check tile versions.

**Comment by joto** [comment]
Having such as tiny thing not-quite-compatible to protocol buffers spec doesn't sound like a good idea to me. Especially if we want this vector tile format to be the "industry standard" of vector tile formats, as it seems to be. Changing this will confuse implementors to no end who think vector tiles use protocol buffers. If we go this route, we might as well just tell everybody, that this is our own special binary format and only our own special libraries can read and write it. That would then open up other possibilites to make encoding fit our use case even better. If I remember correctly we had some other issues with the encoding where we had to use non-optimal encodings because we are restricted to what protocol buffer offers. Or we can move the version number outside the protocol buffers encoded part. If we drop the "simple concatenation" goal, that shouldn't be a problem.

**Comment by TimSC** [comment]
I'm a relative beginner with protobufs but hopefully standard mechanisms can be used if decoding of different layer versions is anything more than an optional feature. I notice OsmAnd has already gone down the non-standard/barely-documented protobuf route. Perhaps have each layer contained entirely within a byte blob, so they can be decoded separately? Alternatively, as @joto suggests, define VectorLayerV2, etc.

**Comment by flippmoke** [comment]
@TimSC we have considered this as an alternative going forward, however, we are not in any hurry to change the standard at this time. Part of the reason it was still done this way for version `2.0` was that we wanted to clarify and make improvements while still being backwards/forward compatible as much as possible.
